File: src/main/java/be/angelcorp/libs/celest/kepler/KeplerParabola.java
/**
 * Copyright (C) 2011 simon <aodtorusan@gmail.com>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *        http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.libs.celest.kepler;

import static java.lang.Math.PI;
import static java.lang.Math.atan2;
import static java.lang.Math.pow;

import org.apache.commons.math.MathRuntimeException;
import org.apache.commons.math.analysis.DifferentiableUnivariateRealFunction;
import org.apache.commons.math.analysis.UnivariateRealFunction;
import org.apache.commons.math.analysis.solvers.UnivariateRealSolver;
import org.apache.commons.math.analysis.solvers.UnivariateRealSolverFactory;

import be.angelcorp.libs.celest.stateVector.IKeplerElements;
import be.angelcorp.libs.math.MathUtils2;

public class KeplerParabola extends KeplerEquations {

	public static double radius(double p, double B) {
		final double r = (p / 2) * (1 + B * B);
		return r;
	}

	public KeplerParabola(IKeplerElements k) {
		super(k);
	}

	@Override
	public double anomalyFromMeanAnomaly(final double M) {
		UnivariateRealFunction f = new DifferentiableUnivariateRealFunction() {

			@Override
			public UnivariateRealFunction derivative() {
				return new UnivariateRealFunction() {
					@Override
					public double value(double B) {
						return 1 + 1 * B * B;
					}
				};
			}

			@Override
			public double value(double B) {
				return B + pow(B, 3) / 3. - M;
			}
		};

		UnivariateRealSolver solver = UnivariateRealSolverFactory.newInstance().newNewtonSolver();
		try {
			final double B = solver.solve(f, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, 0);
			return B;
		} catch (Exception e) {
			throw new MathRuntimeException(e);
		}
	}

	@Override
	public double anomalyFromTrueAnomaly(double nu) {
		nu = MathUtils2.mod(nu, 2 * PI);

		if (nu > PI)
			nu = nu - 2 * PI;

		final double B = StrictMath.tan(nu / 2);
		return B;
	}

	@Override
	public double arealVel(double mu, double a, double e) {
		// TODO Auto-generated method stub
		throw new UnsupportedOperationException();
	}

	@Override
	public double flightPathAngle() {
		final double gamma = k.getTrueAnomaly() / 2;
		return gamma;
	}

	@Override
	public double getApocenter() {
		return Double.POSITIVE_INFINITY;
	}

	@Override
	protected DifferentiableUnivariateRealFunction getFundamentalEquation(double e, double M) {
		// TODO Auto-generated method stub
		throw new UnsupportedOperationException();
	}

	@Override
	public double getPericenter() {
		return semiLatusRectum() / 2;
	}

	@Override
	public double meanAnomalyFromAnomaly(double anomaly) {
		final double M = anomaly + (anomaly * anomaly * anomaly) / 3.0;
		return M;
	}

	@Override
	public double meanAnomalyFromTrue(double nu) {
		return meanAnomalyFromAnomaly(anomaly());
	}

	@Override
	public double period(double n) {
		return Double.POSITIVE_INFINITY;
	}

	@Override
	public double semiLatusRectum() {
		return k.getSemiMajorAxis();// TODO: better solution ?
	}

	@Override
	public double totEnergyPerMass(double mu, double a) {
		return 0;
	}

	@Override
	public double trueAnomalyFromAnomaly(double B) {
		final double p = semiLatusRectum();
		final double r = radius(p, B);
		final double nu = atan2(p * B, (p - r));
		return nu;
	}

	@Override
	public double visViva(double r) {
		return 2 * k.getCenterbody().getMu() / r;
	}

}

File: celest/src/main/java/be/angelcorp/libs/celest/kepler/KeplerParabola.scala
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *        http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.libs.celest.kepler;

import math._
import be.angelcorp.libs.celest.state.positionState.IKeplerElements
import org.apache.commons.math3.analysis.{UnivariateFunction, DifferentiableUnivariateFunction}
import org.apache.commons.math3.analysis.solvers.NewtonSolver
import be.angelcorp.libs.math.MathUtils2

class KeplerParabola(k: IKeplerElements) extends KeplerEquations(k) {

	override def anomalyFromMeanAnomaly(M: Double) = {
		val f = new DifferentiableUnivariateFunction() {
			override def derivative() =
				new UnivariateFunction() {
					override def value(B: Double) = 1 + 1 * B * B
				}

			override def value(B: Double) = B + pow(B, 3) / 3.0 - M
		}

		val solver = new NewtonSolver()
		solver.solve(50, f, M)
	}

	override def anomalyFromTrueAnomaly(trueA: Double) = {
		var nu = MathUtils2.mod(trueA, 2 * Pi)

		if (nu > Pi)
			nu = nu - 2 * Pi

		tan(nu / 2)
	}

	override def arealVel(µ: Double, a: Double, e: Double) = throw new UnsupportedOperationException()

	override def flightPathAngle() = k.getTrueAnomaly/ 2

	override def getApocenter() = Double.PositiveInfinity

	override def getFundamentalEquation(e: Double, M: Double) = throw new UnsupportedOperationException()

	override def getPericenter() = semiLatusRectum() / 2

	override def meanAnomalyFromAnomaly(anomaly: Double) = anomaly + (anomaly * anomaly * anomaly) / 3.0

	override def meanAnomalyFromTrue(nu: Double) = meanAnomalyFromAnomaly(anomaly())

	override def period(n: Double) = Double.PositiveInfinity

	override def semiLatusRectum() = k.getSemiMajorAxis// TODO: better solution ?

	override def totEnergyPerMass(µ: Double, a: Double) = 0

	override def trueAnomalyFromAnomaly(B: Double) = {
		val p = semiLatusRectum()
		val r = KeplerParabola.radius(p, B)
		atan2(p * B, (p - r))
	}

	override def visViva(r: Double) = 2 * k.getCenterbody.getMu / r

}

object KeplerParabola {

	def radius(p: Double, B: Double) = (p / 2) * (1 + B * B)

}
